File: go/tracer.go
package gentrail

import (
	"context"
	"crypto/rand"
	"encoding/hex"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

const (
	invocationSpanName = "governance.invocation"
	modelCallSpanName  = "governance.model_call"
	sourceAttrValue    = "aigentrail-sdk"

	maxInputValueRunes  = 4000
	maxOutputValueRunes = 4000
	maxModelOutputRunes = 4000
)

// Tracer emits governance spans (invocations, model calls, tool calls) over
// OTLP. Construct one per process with New; call Shutdown before exit.
type Tracer struct {
	tracer   trace.Tracer
	provider *sdktrace.TracerProvider
}

// InvocationParams describes a new agent invocation that wraps a sequence of
// model and tool calls.
type InvocationParams struct {
	AgentID     string
	AgentName   string
	JournalID   string
	UserMessage string
}

// InvocationEndParams describes the outcome of an invocation. An empty Status
// is treated as "ok".
type InvocationEndParams struct {
	Response      string
	TotalTokens   int64
	ToolCount     int64
	IntegrityHash string
	Status        string
}

// ModelCallParams describes a single LLM call inside an invocation. A zero
// LatencyMS is omitted from the span.
type ModelCallParams struct {
	ModelID      string
	Prompt       string
	ResponseText string
	InputTokens  int64
	OutputTokens int64
	LatencyMS    float64
}

// ToolCallParams describes a single tool invocation inside an agent
// invocation. A zero DurationMS is omitted from the span.
type ToolCallParams struct {
	AgentID    string
	AgentName  string
	Name       string
	Args       string
	Result     string
	DurationMS float64
}

// LLMCallParams describes a single standalone LLM call for RecordLLMCall.
// JournalID is generated when empty; Status defaults to "ok".
type LLMCallParams struct {
	AgentID      string
	AgentName    string
	JournalID    string
	ModelID      string
	Prompt       string
	Response     string
	InputTokens  int64
	OutputTokens int64
	LatencyMS    float64
	Status       string
}

// Invocation is a handle to an in-flight governance.invocation span. Call End
// exactly once. The context returned by StartInvocation must be passed to
// RecordModelCall and RecordToolCall so they nest correctly.
type Invocation struct {
	span trace.Span
}

// StartInvocation opens a governance.invocation span and returns a context
// scoped to it. Pass that context to RecordModelCall / RecordToolCall so
// child spans nest correctly. JournalID is generated when empty. A nil Tracer
// is a no-op and returns ctx unchanged with a nil Invocation.
func (t *Tracer) StartInvocation(ctx context.Context, p InvocationParams) (context.Context, *Invocation) {
	if t == nil {
		return ctx, nil
	}
	if p.JournalID == "" {
		p.JournalID = newJournalID()
	}
	ctx, span := t.tracer.Start(ctx, invocationSpanName)
	span.SetAttributes(
		attribute.String("openinference.span.kind", "AGENT"),
		attribute.String("aigentrail.agent.id", p.AgentID),
		attribute.String("agent.name", p.AgentName),
		attribute.String("aigentrail.journal.id", p.JournalID),
		attribute.String("session.id", p.JournalID),
		attribute.String("input.value", truncateRunes(p.UserMessage, maxInputValueRunes)),
		attribute.String("source", sourceAttrValue),
	)
	return ctx, &Invocation{span: span}
}

// End closes the invocation span. Status defaults to "ok" if empty. A nil
// Invocation is a no-op.
func (i *Invocation) End(p InvocationEndParams) {
	if i == nil {
		return
	}
	status := p.Status
	if status == "" {
		status = "ok"
	}
	i.span.SetAttributes(
		attribute.String("output.value", truncateRunes(p.Response, maxOutputValueRunes)),
		attribute.String("aigentrail.invocation.status", status),
		attribute.String("aigentrail.journal.integrity_hash", p.IntegrityHash),
		attribute.Int64("llm.token_count.total", p.TotalTokens),
		attribute.Int64("aigentrail.tool.count", p.ToolCount),
	)
	i.span.SetStatus(codes.Ok, "")
	i.span.End()
}

// RecordModelCall emits a governance.model_call span as a child of whatever
// span is in ctx (typically the invocation returned by StartInvocation). A nil
// Tracer is a no-op.
func (t *Tracer) RecordModelCall(ctx context.Context, p ModelCallParams) {
	if t == nil {
		return
	}
	_, span := t.tracer.Start(ctx, modelCallSpanName)
	defer span.End()
	span.SetAttributes(
		attribute.String("openinference.span.kind", "LLM"),
		attribute.String("llm.model_name", p.ModelID),
		attribute.String("input.value", truncateRunes(p.Prompt, maxInputValueRunes)),
		attribute.String("output.value", truncateRunes(p.ResponseText, maxModelOutputRunes)),
		attribute.Int64("llm.token_count.prompt", p.InputTokens),
		attribute.Int64("llm.token_count.completion", p.OutputTokens),
	)
	if p.LatencyMS > 0 {
		span.SetAttributes(attribute.Float64("aigentrail.latency_ms", p.LatencyMS))
	}
}

// RecordToolCall emits a tool-call span (named after the tool) as a child of
// whatever span is in ctx. The agent.id / agent.name attributes let the
// collector re-attach tool calls to their parent invocation even when the
// BatchSpanProcessor flushes children before the parent ends.
func (t *Tracer) RecordToolCall(ctx context.Context, p ToolCallParams) {
	if t == nil {
		return
	}
	_, span := t.tracer.Start(ctx, p.Name)
	defer span.End()
	span.SetAttributes(
		attribute.String("openinference.span.kind", "TOOL"),
		attribute.String("tool.name", p.Name),
		attribute.String("aigentrail.agent.id", p.AgentID),
		attribute.String("agent.name", p.AgentName),
		attribute.String("input.value", truncateRunes(p.Args, maxInputValueRunes)),
		attribute.String("output.value", truncateRunes(p.Result, maxOutputValueRunes)),
	)
	if p.DurationMS > 0 {
		span.SetAttributes(attribute.Float64("aigentrail.latency_ms", p.DurationMS))
	}
}

// RecordLLMCall opens an invocation, records one model call, and ends the
// invocation in a single call. A nil Tracer is a no-op.
func (t *Tracer) RecordLLMCall(ctx context.Context, p LLMCallParams) {
	if t == nil {
		return
	}
	ctx, invocation := t.StartInvocation(ctx, InvocationParams{
		AgentID:     p.AgentID,
		AgentName:   p.AgentName,
		JournalID:   p.JournalID,
		UserMessage: p.Prompt,
	})
	t.RecordModelCall(ctx, ModelCallParams{
		ModelID:      p.ModelID,
		Prompt:       p.Prompt,
		ResponseText: p.Response,
		InputTokens:  p.InputTokens,
		OutputTokens: p.OutputTokens,
		LatencyMS:    p.LatencyMS,
	})
	invocation.End(InvocationEndParams{
		Response:    p.Response,
		TotalTokens: p.InputTokens + p.OutputTokens,
		Status:      p.Status,
	})
}

func newJournalID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return ""
	}
	return hex.EncodeToString(b[:])
}

func truncateRunes(s string, n int) string {
	if n <= 0 {
		return ""
	}
	runes := []rune(s)
	if len(runes) <= n {
		return s
	}
	return string(runes[:n])
}

File: js/gentrail-ai/src/tracer.ts
import { randomUUID } from "node:crypto";

import {
  context,
  SpanStatusCode,
  trace,
  type Span,
  type Tracer,
  type TracerProvider,
} from "@opentelemetry/api";
import { BasicTracerProvider, BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";

import {
  buildGentrailOtlpExporter,
  gentrailApiKeyFromEnv,
  redactionEnabledFromEnv,
} from "./otlpConfig.js";
import { redactPII } from "./pii.js";

export const GOVERNANCE_SOURCE = "aigentrail-sdk";
export const VALUE_CODE_POINTS_MAX = 4000;
export const INVOCATION_SPAN_NAME = "governance.invocation";
export const MODEL_CALL_SPAN_NAME = "governance.model_call";
const TRACER_NAME = "aigentrail.governance";

export interface FlushableTracerProvider extends TracerProvider {
  forceFlush(): Promise<void>;
  shutdown(): Promise<void>;
}

export interface GovernanceTracerOptions {
  tracerProvider: FlushableTracerProvider;
  redact?: boolean;
}

export interface InvocationStart {
  agentId: string;
  agentName: string;
  journalId: string;
  userMessage: string;
}

export interface InvocationEnd {
  response: string;
  totalTokens: number;
  toolCount: number;
  integrityHash: string;
  status?: string;
}

export interface ModelCall {
  modelId: string;
  prompt: string;
  responseText: string;
  inputTokens: number;
  outputTokens: number;
  latencyMs?: number;
}

export interface ToolCall {
  agentId: string;
  agentName: string;
  name: string;
  args: string;
  result: string;
  durationMs?: number;
  enforcedDecision?: string;
}

export interface LLMCall {
  agentId?: string;
  agentName?: string;
  journalId?: string;
  modelId: string;
  prompt: string;
  responseText: string;
  inputTokens?: number;
  outputTokens?: number;
  totalTokens?: number;
  latencyMs?: number;
  status?: string;
}

export interface InvocationHandle {
  readonly span: Span;
}

export type FetchFunction = (
  input: string | URL | Request,
  init?: RequestInit,
) => Promise<Response>;

export class GovernanceTracer {
  private readonly tracer: Tracer;
  private readonly tracerProvider: FlushableTracerProvider;
  private readonly redact: boolean;

  constructor(options: GovernanceTracerOptions) {
    this.tracerProvider = options.tracerProvider;
    this.tracer = options.tracerProvider.getTracer(TRACER_NAME);
    this.redact = options.redact ?? true;
  }

  startInvocation(start: InvocationStart): InvocationHandle {
    const span = this.tracer.startSpan(INVOCATION_SPAN_NAME, {}, context.active());
    span.setAttributes({
      "openinference.span.kind": "AGENT",
      "aigentrail.agent.id": start.agentId,
      "agent.name": start.agentName,
      "aigentrail.journal.id": start.journalId,
      "session.id": start.journalId,
      "input.value": this.attributeValue(start.userMessage),
      source: GOVERNANCE_SOURCE,
    });
    return { span };
  }

  endInvocation(handle: InvocationHandle, end: InvocationEnd): void {
    handle.span.setAttributes({
      "output.value": this.attributeValue(end.response),
      "aigentrail.invocation.status": end.status ?? "ok",
      "aigentrail.journal.integrity_hash": end.integrityHash,
      "llm.token_count.total": end.totalTokens,
      "aigentrail.tool.count": end.toolCount,
    });
    handle.span.setStatus({ code: SpanStatusCode.OK });
    handle.span.end();
  }

  recordModelCall(parent: InvocationHandle, call: ModelCall): void {
    const span = this.tracer.startSpan(
      MODEL_CALL_SPAN_NAME,
      {},
      trace.setSpan(context.active(), parent.span),
    );
    span.setAttributes({
      "openinference.span.kind": "LLM",
      "llm.model_name": call.modelId,
      "input.value": this.attributeValue(call.prompt),
      "output.value": this.attributeValue(call.responseText),
      "llm.token_count.prompt": call.inputTokens,
      "llm.token_count.completion": call.outputTokens,
    });
    if (call.latencyMs !== undefined) {
      span.setAttribute("aigentrail.latency_ms", call.latencyMs);
    }
    span.end();
  }

  recordToolCall(parent: InvocationHandle, call: ToolCall): void {
    const span = this.tracer.startSpan(call.name, {}, trace.setSpan(context.active(), parent.span));
    span.setAttributes({
      "openinference.span.kind": "TOOL",
      "tool.name": call.name,
      "aigentrail.agent.id": call.agentId,
      "agent.name": call.agentName,
      "input.value": this.attributeValue(call.args),
      "output.value": this.attributeValue(call.result),
    });
    if (call.durationMs !== undefined) {
      span.setAttribute("aigentrail.latency_ms", call.durationMs);
    }
    if (call.enforcedDecision) {
      span.setAttribute("aigentrail.enforcement.decision", call.enforcedDecision);
    }
    span.end();
  }

  recordLLMCall(call: LLMCall): void {
    const inputTokens = call.inputTokens ?? 0;
    const outputTokens = call.outputTokens ?? 0;
    const parent = this.startInvocation({
      agentId: call.agentId ?? "",
      agentName: call.agentName ?? "",
      journalId: call.journalId || randomUUID().replaceAll("-", ""),
      userMessage: call.prompt,
    });
    this.recordModelCall(parent, {
      modelId: call.modelId,
      prompt: call.prompt,
      responseText: call.responseText,
      inputTokens,
      outputTokens,
      latencyMs: call.latencyMs,
    });
    this.endInvocation(parent, {
      response: call.responseText,
      totalTokens: call.totalTokens ?? inputTokens + outputTokens,
      toolCount: 0,
      integrityHash: "",
      status: call.status,
    });
  }

  fetch(baseFetch: FetchFunction = globalThis.fetch): FetchFunction {
    return async (input, init) => {
      const startedAt = performance.now();
      const request = describeRequest(input, init);
      let status = "ok";
      try {
        const response = await baseFetch(input, init);
        if (response.status >= 400) {
          status = `http_${response.status}`;
        }
        return response;
      } catch (err) {
        status = "error";
        throw err;
      } finally {
        this.recordLLMCall({
          modelId: request.host,
          prompt: `${request.method} ${request.pathname}`,
          responseText: status,
          latencyMs: Math.floor(performance.now() - startedAt),
          status,
        });
      }
    };
  }

  forceFlush(): Promise<void> {
    return this.tracerProvider.forceFlush();
  }

  shutdown(): Promise<void> {
    return this.tracerProvider.shutdown();
  }

  private attributeValue(value: string): string {
    return truncateCodePoints(this.redact ? redactPII(value) : value, VALUE_CODE_POINTS_MAX);
  }
}

interface RequestDescription {
  method: string;
  host: string;
  pathname: string;
}

function describeRequest(
  input: string | URL | Request,
  init: RequestInit | undefined,
): RequestDescription {
  const isRequest = input instanceof Request;
  const rawUrl = isRequest ? input.url : String(input);
  const method = (init?.method ?? (isRequest ? input.method : "GET")).toUpperCase();
  if (!URL.canParse(rawUrl)) {
    return { method, host: "", pathname: rawUrl };
  }
  const url = new URL(rawUrl);
  return { method, host: url.host, pathname: url.pathname };
}

export function truncateCodePoints(value: string, codePointsMax: number): string {
  if (value.length <= codePointsMax) {
    return value;
  }
  let codePoints = 0;
  let index = 0;
  while (index < value.length && codePoints < codePointsMax) {
    const codePoint = value.codePointAt(index) ?? 0;
    index += codePoint > 0xffff ? 2 : 1;
    codePoints += 1;
  }
  return value.slice(0, index);
}

export interface CreateGovernanceTracerOptions {
  redact?: boolean;
}

export function createGovernanceTracer(
  options: CreateGovernanceTracerOptions = {},
): GovernanceTracer | null {
  const apiKey = gentrailApiKeyFromEnv();
  if (apiKey === "") {
    return null;
  }
  const exporter = buildGentrailOtlpExporter(apiKey);
  const tracerProvider = new BasicTracerProvider({
    spanProcessors: [new BatchSpanProcessor(exporter)],
  });
  return new GovernanceTracer({
    tracerProvider,
    redact: options.redact ?? redactionEnabledFromEnv(),
  });
}
